controllers/managedocs_controller.go

```diff
@@ -710,67 +706,10 @@ func (r *ManagedOCSReconciler) reconcileAlertmanagerConfigSecret() error {
-		Details    map[string]string `yaml:"details"`
-	}
-
-	type WebhookConfig struct {
-		Url string `yaml:"url"`
-	}
-
-	alertmanagerConfig := struct {
-		Route struct {
-			GroupWait      string `yaml:"group_wait,omitempty"`
-			GroupInterval  string `yaml:"group_interval,omitempty"`
-			RepeatInterval string `yaml:"repeat_interval,omitempty"`
-			Receiver       string `yaml:"receiver,omitempty"`
-			Routes         [2]struct {
-				GroupWait      string            `yaml:"group_wait,omitempty"`
-				GroupInterval  string            `yaml:"group_interval,omitempty"`
-				RepeatInterval string            `yaml:"repeat_interval,omitempty"`
-				Receiver       string            `yaml:"receiver,omitempty"`
-				GroupBy        []string          `yaml:"group_by,omitempty"`
-				Match          map[string]string `yaml:"match,omitempty"`
-			} `yaml:"routes"`
-		} `yaml:"route"`
-		Receivers [2]struct {
-			Name             string            `yaml:"name"`
-			PagerdutyConfigs []PagerDutyConfig `yaml:"pagerduty_configs,omitempty"`
-			WebhookConfigs   []WebhookConfig   `yaml:"webhook_configs,omitempty"`
-		} `yaml:"receivers"`
-	}{}
-
-	alertmanagerConfig.Route.Receiver = "pagerduty"
-
-	alertmanagerConfig.Route.Routes[0].GroupWait = "30s"
-	alertmanagerConfig.Route.Routes[0].GroupInterval = "5m"
-	alertmanagerConfig.Route.Routes[0].RepeatInterval = "4h"
-	alertmanagerConfig.Route.Routes[0].Receiver = "pagerduty"
-	alertmanagerConfig.Route.Routes[0].Match = make(map[string]string)
-	alertmanagerConfig.Route.Routes[0].Match["severity"] = "critical"
-	alertmanagerConfig.Route.Routes[0].GroupBy = []string{"alertname"}
-
-	alertmanagerConfig.Route.Routes[1].GroupWait = "5m"
-	alertmanagerConfig.Route.Routes[1].GroupInterval = "5m"
-	alertmanagerConfig.Route.Routes[1].RepeatInterval = "5m"
-	alertmanagerConfig.Route.Routes[1].Receiver = "DeadMansSnitch"
-	alertmanagerConfig.Route.Routes[1].Match = make(map[string]string)
-	alertmanagerConfig.Route.Routes[1].Match["alertname"] = "DeadMansSnitch"
-
-	alertmanagerConfig.Receivers[0].Name = "pagerduty"
-	alertmanagerConfig.Receivers[0].PagerdutyConfigs = []PagerDutyConfig{{}}
-	alertmanagerConfig.Receivers[0].PagerdutyConfigs[0].ServiceKey = pagerdutyServiceKey
-	alertmanagerConfig.Receivers[0].PagerdutyConfigs[0].Details = map[string]string{
-		"SOP": r.SOPEndpoint,
-	}
-
-	alertmanagerConfig.Receivers[1].Name = "DeadMansSnitch"
-	alertmanagerConfig.Receivers[1].WebhookConfigs = []WebhookConfig{{}}
-	alertmanagerConfig.Receivers[1].WebhookConfigs[0].Url = dmsURL
-
-	return &alertmanagerConfig
+func (r *ManagedOCSReconciler) updateAlertmanagerConfig(ac *promv1a1.AlertmanagerConfig, pagerdutyServiceKey string, dmsURL string) {
+	ac.Spec.Receivers[0].PagerDutyConfigs[0].ServiceKey = &corev1.SecretKeySelector{Key: pagerdutyServiceKey}
+	ac.Spec.Receivers[0].PagerDutyConfigs[0].Details = []promv1a1.KeyValue{{Key: "SOP"}, {Value: r.SOPEndpoint}}
+	ac.Spec.Receivers[1].WebhookConfigs[0].URL = &dmsURL
```

Comment:
Assuming that a specific receiver is located at a specific index is problematic, the code will break if we add another receiver in the future as the first reciver. 

To mitigate you can just search for the correct receiver, as an example:

```suggestion
for i := range ac.Spec.Receivers {
    receiver := &range ac.Spec.Receivers[i]
    switch receiver.name {
    case "pagerduty":
        receiver.PagerDutyConfigs[0].ServiceKey.Key = pagerdutyServiceKey
        receiver.PagerDutyConfigs[0].Details[0].Key = "SOP"
        receiver.PagerDutyConfigs[0].Details[0].Value = r.SOPEndpoint
    case "DeadManSnitch":
        receiver.WebhookConfigs[0].URL = &dmsURL
    }
}
```

Please note the following (In my suggestion above):
1. The structure itself (arrays, selector, etc.) should move to the template. No reason to create it in code.
2. It is ok to directly access via an index a config for a specific receiver as (e.g. `PagerDutyConfigs[0]`) as this is internal structure for that specific receiver